File: src/test/java/im/tox/tox4j/ToxAvImplTestBase.java
package im.tox.tox4j;

import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.av.ToxAv;
import im.tox.tox4j.av.ToxAvTestBase;
import im.tox.tox4j.av.exceptions.ToxAvNewException;
import im.tox.tox4j.core.ToxCore;
import im.tox.tox4j.core.ToxOptions;
import im.tox.tox4j.core.exceptions.ToxNewException;
import org.junit.After;

import java.util.ArrayList;
import java.util.Collection;

public abstract class ToxAvImplTestBase extends ToxAvTestBase {

    private static final DhtNode node = new DhtNodeSelector().node(new ToxFactory() {
        @NotNull
        @Override
        public ToxCore newTox(boolean ipv6Enabled, boolean udpEnabled) throws ToxNewException {
            ToxOptions options = new ToxOptions();
            options.setIpv6Enabled(ipv6Enabled);
            options.setUdpEnabled(udpEnabled);
            return new ToxCoreImpl(options);
        }
    });

    private final Collection<ToxCoreImpl> toxes = new ArrayList<>();
    private final Collection<ToxAvImpl> avs = new ArrayList<>();

    @After
    public void tearDown() {
        for (ToxAvImpl av : avs) {
            av.close();
        }
        avs.clear();
        for (ToxCoreImpl tox : toxes) {
            tox.close();
        }
        toxes.clear();
        System.gc();
    }

    @NotNull
    @Override
    protected final ToxCore newTox(ToxOptions options, byte[] data) throws ToxNewException {
        ToxCoreImpl tox = new ToxCoreImpl(options, data);
        toxes.add(tox);
        return tox;
    }

    @Override
    protected final ToxAv newToxAv(ToxCore tox) throws ToxAvNewException {
        ToxAvImpl av = new ToxAvImpl(tox);
        avs.add(av);
        return av;
    }

    @NotNull
    @Override
    protected DhtNode node() {
        return node;
    }

}

File: src/test/java/im/tox/tox4j/ToxAvImplTestBase.scala
package im.tox.tox4j

import im.tox.tox4j.annotations.NotNull
import im.tox.tox4j.av.exceptions.ToxAvNewException
import im.tox.tox4j.av.{ ToxAv, ToxAvTestBase }
import im.tox.tox4j.core.ToxCore
import im.tox.tox4j.core.exceptions.ToxNewException
import im.tox.tox4j.core.options.ToxOptions
import im.tox.tox4j.impl.{ ToxAvImpl, ToxCoreImpl }
import org.junit.After

import scala.collection.mutable.ArrayBuffer

abstract class ToxAvImplTestBase extends ToxAvTestBase {

  protected val node: DhtNode = DhtNodeSelector.node

  private final val toxes = new ArrayBuffer[ToxCoreImpl]
  private final val avs = new ArrayBuffer[ToxAvImpl]

  @After
  def tearDown(): Unit = {
    avs.foreach(_.close())
    avs.clear()
    toxes.foreach(_.close())
    toxes.clear()
    System.gc()
  }

  @NotNull
  @throws[ToxNewException]
  protected final def newTox(options: ToxOptions): ToxCore = {
    val tox = new ToxCoreImpl(options)
    toxes += tox
    tox
  }

  @throws[ToxAvNewException]
  protected final def newToxAv(tox: ToxCore): ToxAv = {
    val av = new ToxAvImpl(tox.asInstanceOf[ToxCoreImpl])
    avs += av
    av
  }

}
